hw06 - annotations, reflection

qwelle opened the pull request.

qwelle committed:
hw06 - annotations, reflection

petrelevich commented on hw06-reflections/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    static void run(Class<?> clazz) throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
+        ArrayList<String> arrTests = new ArrayList<>();
```
не забывайте private final
и лучше 
List<String> arrTests = new ArrayList<>();

qwelle replied:
это же переменные.
Методы сделала private

petrelevich replied:
да, действительно.
извините, не туда посмотрел.

petrelevich commented on hw06-reflections/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,74 @@
+            String methodName = method.getName();
+            Annotation[] annotations = method.getDeclaredAnnotations();
+            for (Annotation annotation : annotations) {
+                switch (annotation.annotationType().getSimpleName()) {
```
Посмотрите
isAnnotationPresent()
https://www.tutorialspoint.com/java/lang/package_isannotationpresent.htm

qwelle replied:
ок, поправила

petrelevich commented on hw06-reflections/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,74 @@
+                        arrBefore.add(methodName);
+                        break;
+                    case ("After"):
+                        arrAfter.add(methodName);
```
эта и другие коллекции используются локально.
зачем из выносить в поля?

qwelle replied:
у меня не было полей, все переменные в одном статическом методе

petrelevich commented on hw06-reflections/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,74 @@
+
+        for (String testMethod : arrTests) {
+            System.out.println("------------------------------------");
+            //��� ������� ������ ���� ������
```
видимо windows-кодировка?
в в веб-интерфейсе github не читается.

qwelle replied:
поправила

petrelevich commented on hw06-reflections/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,74 @@
+
+        Method[] methods = clazz.getDeclaredMethods();
+
+        for(Method method : methods) {
```
длинный метод лучше разбить на несколько:
подготовка, выполение, вывод результата

И попробуйте обойтись без полей класса.

qwelle replied:
разбила на три метода, но получила неудобство - до 5 формальных параметров у методов, что можно придумать для лаконичности и при этом чтобы не было полей??

petrelevich replied:
методы не надо делать статическими, а передавать можно экземпляры классов (см. коммент выше)

qwelle committed:
Устранение замечаний:
декомпозиция на private методы
isAnnotationPresent() вместо getDeclaredAnnotations()
List arrTests = new ArrayList<>() вместо ArrayList<String> arrTests = new ArrayList<>();
исправлена кодировка

petrelevich commented on hw06-reflections/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,73 @@
+public class TestRunner {
+
+    public static void main(String[] args) throws InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
+        List arrTests = new ArrayList<>();
```
не забывайте про дженерики

petrelevich commented on hw06-reflections/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,73 @@
+        List arrTests = new ArrayList<>();
+        List arrBefore = new ArrayList<>();
+        List arrAfter = new ArrayList<>();
+        List arrSuccess = new ArrayList<>();
```
эти переменные можно объединить в два класса: "тесты" и "результаты тестов" и передамать между методами экземпляры этих классов.

petrelevich commented on hw06-reflections/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,73 @@
+        }
+    }
+
+    private static void runTests(Class<?> clazz, List arrBefore, List arrAfter, List arrTests, List arrSuccess, List arrFailed) throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
```
не забывайте про дженерики

qwelle committed:
вынос переменных в классы
меоды нестатические

petrelevich commented on hw06-reflections/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,82 @@
+    List<String> arrAfter = new ArrayList<>();
+}
+
+class TestsResults {
```
эти классы стоит сделать полноценными: с приватными полями и т.д.

petrelevich commented on hw06-reflections/src/main/java/homework/TestRunner.java:
```diff
@@ -0,0 +1,82 @@
+        runner.printResult(testArrays, resultArrays);
+    }
+
+    private void prepareData(Class<?> clazz, Tests testArrays) {
```
лучше, чтобы метод создал и вернул testArrays, а не менял переданную сущность.

qwelle committed:
Merge branch 'main' into hw06-reflections
